### ISSUE TITLE
Can FP8 GEMM be enabled via module hooks instead of module swapping?

### ISSUE THREAD

**Issue by zigzagcai** [pytorch/ao#1887]
Hi developers,

Thanks for such a great project!

I want to integrate torchao FP8 GEMM into our training framework. But in my framework, the linear layers are defined in customized modules (where we implement Tensor Parallel or ZeRO3 weight parallel), so it is hard to directly swap the linear layers with torchao `Float8Linear`.

So, can FP8 GEMM enabled via a more friendly way, such like module hooks? Since module swapping is not so flexible

**Comment by danielvegamyhre** [comment]
For float8 training, module swap is currently the only supported method in torchao.

However, you could implement a module hook which uses our quantization primitives to convert high precision tensors to Float8Tensors, and then use torch._scaled_mm for the float8 gemm (which [dispatches to cuBLAS](https://github.com/pytorch/pytorch/blob/f2ea77c09956b825f4cf5acf08917e523627159a/aten/src/ATen/native/cuda/Blas.cpp#L1339) for tensorwise scaling and [dispatches to cutlass](https://github.com/pytorch/pytorch/blob/f2ea77c09956b825f4cf5acf08917e523627159a/aten/src/ATen/native/cuda/Blas.cpp#L1200-L1202) for rowwise scaling).

We could potentially provide such a hook in torchao, though it is not currently planned, I think it could be useful though for use cases like this. @vkuzo @drisspg any thoughts on this?

**Comment by vkuzo** [comment]
We started with module swapping because it's the easiest way to iterate on performance/accuracy/usability.  If we can cover more important use cases with alternate UX options, I'm in favor.

@zigzagcai , a couple of questions for you.
1. are you using torch.compile / are you open to using torch.compile
2. could you share a pointer to your code if you have it

From what you have shared so far, a tensor subclass weight wrapper sounds like the right UX, but would be good to see the callsites to confirm that.

**Comment by vkuzo** [comment]
thanks, @zigzagcai .  I think a tensor subclass weight wrapper is promising for your use case.

We have a prototype feature for quantized training with int8 with a tensor subclass weight wrapping UX here: https://github.com/pytorch/ao/tree/main/torchao/prototype/quantized_training#int8-mixed-precision-training . Would you be up for trying this and seeing if it works with your use case?  If yes, that would be great signal for adding an option for this UX for float8.

**Comment by zigzagcai** [comment]
> thanks, @zigzagcai .  I think a tensor subclass weight wrapper is promising for your use case.
> 
> We have a prototype feature for quantized training with int8 with a tensor subclass weight wrapping UX here: https://github.com/pytorch/ao/tree/main/torchao/prototype/quantized_training#int8-mixed-precision-training . Would you be up for trying this and seeing if it works with your use case?  If yes, that would be great signal for adding an option for this UX for float8.

Thanks @vkuzo 
Could you please share a pointer to the tensor subclass wrapper?
And I see it is for INT8 training,so how could it be adjusted to FP8 training.

**Comment by vkuzo** [comment]
> Thanks @vkuzo

> Could you please share a pointer to the tensor subclass wrapper?

Yes, when you do the following

```python
quantize_(model, int8_mixed_precision_training())
```

Then the weights of `torch.nn.Linear` modules will be swapped with `Int8MixedPrecisionTrainingLinearWeight` (code: https://github.com/pytorch/ao/blob/main/torchao/prototype/quantized_training/int8_mixed_precision.py#L300C22-L300C60).  You may need to adjust `filter_fn` to apply this to your custom linear modules, something like

```python
quantize_(model, int8_mixed_precision_training(), filter_fn=lambda mod, fqn: return isinstance(mod, YourCustomModule))
```

> And I see it is for INT8 training,so how could it be adjusted to FP8 training.

Oh, I'm just asking if you're up for trying the int8 training feature to see if it already works for the way you set up your custom linear modules.  If it does, that will make it easy for us to see if we can add an equivalent float8 UX in the future.  If not, we'd love to learn what didn't work, which would help us brainstorm how we can solve your issue.

**Comment by zigzagcai** [comment]
> > Thanks [@vkuzo](https://github.com/vkuzo)
> 
> > Could you please share a pointer to the tensor subclass wrapper?
> 
> Yes, when you do the following
> 
> quantize_(model, int8_mixed_precision_training())
> Then the weights of `torch.nn.Linear` modules will be swapped with `Int8MixedPrecisionTrainingLinearWeight` (code: https://github.com/pytorch/ao/blob/main/torchao/prototype/quantized_training/int8_mixed_precision.py#L300C22-L300C60). You may need to adjust `filter_fn` to apply this to your custom linear modules, something like
> 
> quantize_(model, int8_mixed_precision_training(), filter_fn=lambda mod, fqn: return isinstance(mod, YourCustomModule))
> > And I see it is for INT8 training,so how could it be adjusted to FP8 training.
> 
> Oh, I'm just asking if you're up for trying the int8 training feature to see if it already works for the way you set up your custom linear modules. If it does, that will make it easy for us to see if we can add an equivalent float8 UX in the future. If not, we'd love to learn what didn't work, which would help us brainstorm how we can solve your issue.

Thank you @vkuzo !
I will give it a try!

**Comment by jerryzh168** [comment]
@zigzagcai is this resolved now?

**Comment by zigzagcai** [comment]
> [@zigzagcai](https://github.com/zigzagcai) is this resolved now?

Hi @jerryzh168 , thank you for the follow-up!

Sorry, I was busy with other things previously. Could you keep this issue not closed? And I will try the INT8 training recipe ASAP.
